provider_gsuite.go

```diff
@@ -0,0 +1,216 @@
+package jwtauth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/mitchellh/mapstructure"
+	"golang.org/x/oauth2/google"
+	"golang.org/x/oauth2/jwt"
+	admin "google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/option"
+)
+
+// GSuiteProvider provides G Suite-specific configuration and behavior.
+type GSuiteProvider struct {
+	config    GSuiteProviderConfig // Configuration for the provider
+	jwtConfig *jwt.Config          // Google JWT configuration
+	adminSvc  *admin.Service       // Google admin service
+}
+
+// GSuiteProviderConfig represents the configuration for a GSuiteProvider.
+type GSuiteProviderConfig struct {
+	// Path to a Google service account key file. Required.
+	ServiceAccountFilePath string `mapstructure:"gsuite_service_account"`
+
+	// Email address of a G Suite admin to impersonate. Required.
+	AdminImpersonateEmail string `mapstructure:"gsuite_admin_impersonate"`
+
+	// If set to true, groups will be fetched from G Suite.
+	FetchGroups bool `mapstructure:"fetch_groups"`
+
+	// If set to true, user info will be fetched from G Suite using UserCustomSchemas.
+	FetchUserInfo bool `mapstructure:"fetch_user_info"`
+
+	// Group membership recursion max depth (0 = do not recurse).
+	GroupsRecurseMaxDepth int `mapstructure:"groups_recurse_max_depth"`
+
+	// Comma-separated list of G Suite custom schemas to fetch as claims.
+	UserCustomSchemas string `mapstructure:"user_custom_schemas"`
+
+	// JSON contents of a Google service account key file.
+	serviceAccountKeyJSON []byte
+}
+
+// Initialize initializes the GSuiteProvider by validating and creating configuration.
+func (g *GSuiteProvider) Initialize(jc *jwtConfig) error {
+	// Decode the provider config
+	var config GSuiteProviderConfig
+	if err := mapstructure.Decode(jc.ProviderConfig, &config); err != nil {
+		return err
+	}
+
+	// Read the Google service account key file
+	keyJSON, err := ioutil.ReadFile(config.ServiceAccountFilePath)
+	if err != nil {
+		return err
+	}
+	config.serviceAccountKeyJSON = keyJSON
+
+	return g.initialize(config)
+}
+
+func (g *GSuiteProvider) initialize(config GSuiteProviderConfig) error {
+	var err error
+
+	// Validate configuration
+	if config.ServiceAccountFilePath == "" {
+		return errors.New("'gsuite_service_account' must be set to the file path for a " +
+			"service account key")
+	}
+	if config.AdminImpersonateEmail == "" {
+		return errors.New("'gsuite_admin_impersonate' must be set to an email address of a " +
+			"G Suite user with permission to access the Admin APIs")
```

Comment:
I'd suggest updating this to be more specific to the permissions needed. i.e
```
"'gsuite_admin_impersonate' must be set to an email address of a " +
			"GSuite user with 'Read'  permission to access the GSuite Admin User and Group APIs"
```

Reply:
Good suggestion. Done in eb41bc981111c0bdf0edfd6d4696909f1cab9a94.